### PR TITLE
[inductor] Constant folding for dynamic shape node before pattern matching

### LINKED ISSUE #128387: [Inductor] Constant folding for dynamic shape `full` node before pattern matching
### 🚀 The feature, motivation and pitch

Hi, I'm having a problem with constant folding in dynamic shape. 
Here is the code for a simple reproduction of this situation:
```python
import torch

def foo(x: torch.Tensor):
    batch, length = x.shape
    mask = torch.ones((1, length), device=x.device, dtype=x.dtype)
    mask = 1 - mask
    tmp = x + mask
    return tmp * 2

x = torch.randn((2, 4))
with torch.no_grad():
    torch.compile(foo, dynamic=True)(x)
```
Here is the graph after `constant_fold_uniform_value` in `joint_graph_passes`:
```python
#  arg0_1: "Sym(s0)", arg1_1: "Sym(s1)", arg2_1: "f32[s0, s1]"
graph():
    %arg0_1 : [num_users=0] = placeholder[target=arg0_1]
    %arg1_1 : [num_users=1] = placeholder[target=arg1_1]
    %arg2_1 : [num_users=1] = placeholder[target=arg2_1]
    %full : [num_users=1] = call_function[target=torch.ops.aten.full.default](args = ([1, %arg1_1], 1), kwargs = {dtype: torch.float32, layout: torch.strided, device: cpu, pin_memory: False})
    %sub : [num_users=1] = call_function[target=torch.ops.aten.sub.Tensor](args = (1, %full), kwargs = {})
    %add : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%arg2_1, %sub), kwargs = {})
    %mul : [num_users=1] = call_function[target=torch.ops.aten.mul.Tensor](args = (%add, 2), kwargs = {})
    return (mul,)
```
If close dynamic shape, the graph is:
```python
graph():
    %arg0_1 : [num_users=1] = placeholder[target=arg0_1]
    %full : [num_users=0] = call_function[target=torch.ops.aten.full.default](args = ([1, 4], 1), kwargs = {dtype: torch.float32, layout: torch.strided, device: cpu, pin_memory: False})
    %full_default : [num_users=0] = call_function[target=torch.ops.aten.full.default](args = ([1, 4], 0.0), kwargs = {dtype: torch.float32, layout: torch.strided, device: cpu, pin_memory: False})
    %mul : [num_users=1] = call_function[target=torch.ops.aten.mul.Tensor](args = (%arg0_1, 2), kwargs = {})
    return (mul,)
```
Replacing sub node as full_default, and folding add node.
Such folding cannot be done under dynamic shape,  but logically, we can fold mask tensor. By the way, I think we also can remove two full nodes in final graph (not the focus of this issue).

This improvement will help a lot of sdpa calls to flash attention(no mask) instead of memory efficient attention(with mask), thereby significantly improving performance.

I have seen this PR https://github.com/pytorch/pytorch/pull/118411 (for unbacked symint) and would like to ask for your opinions @ezyang @eellison. Is it possible to do constant folding with hint size(specialize full and save the value).

### Alternatives

_No response_

### Additional context

_No response_

cc @ezyang @bdhirsh @anijain2305 @chauhang

### PR DESCRIPTION
**Pull request by imzhuhl**
Extend constant folding for dynamic shape node, only support pointwise op and some restricted ops 

We support dynamic shapes by limiting constant folding of ops that are guaranteed to have uniform values (`full`, pointwise ops, and views) and running these operators with tensors of shape 1. This also eliminates the possibility of memory overhead of constant folding.

Test plan:

```
python test/inductor/test_torchinductor_dynamic_shapes.py -k test_constant_fold_uniform_value_dynamic
```

Fixes #128387

@eellison 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @ezyang @bdhirsh @anijain2305

### CODE REVIEW

**Review by eellison — COMMENTED** [review]
looking good ! we just got a request to reduce memory usage from Uniform Value constant folder. i know this is not exactly what was originally scoped but it would be great to fold the two changes together since I think they are similar.

I'm thinking along the lines of:

- add a mechanism for custom running of node in ConstantFold

- Within ConstantFoldUniformValue - For non constructors, non-pointwise nodes, non-views return `unknown_value` .

- Otherwise:

For constructors, always construct a tensor of size [1]
For views, just return the input
Pointwise nodes can just defer to the normal implementation. 

Then, as you are doing now, when we insert nodes at the end just take the size from meta. let me know what you think

**Comment by imzhuhl** [comment]
> reduce memory usage from Uniform Value constant folder.

Do you mean to expect to reduce memory usage under both static and dynamic shapes?
You say “For non constructors, non-pointwise nodes, non-views return unknown_value” and “For constructors, always construct a tensor of size [1]”, I realize that these can reduce memory usage in both static and dynamic shape cases, so I have the above questions.

What I understand is to reduce memory usage by reducing the number of running nodes(early return) or the tensor size created by running nodes, is it accurate?

Also, I would like to ask if there is a way to quickly know which ops are constructors or views? (I currently know pointwise ops via torch.Tag.pointwise).

File: torch/_inductor/config.py [hunk 1]
```diff
@@ -394,6 +394,9 @@ def fx_graph_remote_cache_default():
 # constant folding on the joint graph
 joint_graph_constant_folding = True
 
+# support dynamic shape derivation in constant_fold_uniform_value
+enable_dynamic_constant_fold_uniform_value = False
```
[the remark this answers is not in the document]
Comment (imzhuhl) [reply torch/_inductor/config.py:398]: ok

**Comment by eellison** [comment]
> Do you mean to expect to reduce memory usage under both static and dynamic shapes?

Yes, this is to reduce memory overhead of constant folding for both cases.

> Also, I would like to ask if there is a way to quickly know which ops are constructors or views? (I currently know pointwise ops via torch.Tag.pointwise). 

There's not :( See https://github.com/pytorch/pytorch/issues/129020


But i think at this point, in the joint graph, all of the constructors will have been decomposed to `torch.ops.aten.full.default`. 

For views you can either use ops in partioners.py https://github.com/pytorch/pytorch/blob/main/torch/_functorch/partitioners.py#L1194-L1203 or check that the storage on input and output of op on the fake tensor meta is the same

**Comment by imzhuhl** [comment]
hi @eellison, I am not clear with what constant_data_ptr_count does. Can you explain it? I have temporarily delete it.
I think this commit completes the idea of ​​reduce memory, please review it.

**Review by eellison — COMMENTED** [review]
Looks great, thanks for making changes! 

Two things:

1 - don't think we are handling interaction of `x * sym_shape` right now

2 - Reuse `_SymHashingDict` to go from symint -> fx node

**Comment by eellison** [comment]
cc @imzhuhl are you able to update this ? If not, I can take to finish line

**Comment by imzhuhl** [comment]
> cc @imzhuhl are you able to update this ? If not, I can take to finish line

You can finish this, I haven't started yet in fact.

**Review by eellison — APPROVED** [review]

**Comment by eellison** [comment]
Replaced by https://github.com/pytorch/pytorch/pull/129686
